File: demo_parallel_line_fitting.py
#!/usr/bin/env python
import time
import numpy as np
import matplotlib.pyplot as plt
from parallel_fitting.ransac import run_ransac
from parallel_fitting.parallel_line_fitting import estimate, is_inlier


def create_line_points(start, direction, t_range=np.arange(0, 20, 0.5), noise_range=[0, 0.5], random_seed=2021):
    assert start.shape == (1, 2)
    assert direction.shape == (1, 2)
    np.random.seed(random_seed)

    range_arr = np.hstack((t_range[:, np.newaxis], t_range[:, np.newaxis]))
    points = start + direction * range_arr
    noise = np.random.normal(*noise_range, points.shape)
    return points + noise


def create_data():
    direction = np.array([[3.2, 6.7]])
    direction /= np.linalg.norm(direction) + 1e-10

    xys_list = [None] * 2
    xys_list[0] = create_line_points(
        start=np.array([[20.0, 10.0]]),
        direction=direction,
        t_range=np.arange(0, 100, 1.2),
        noise_range=[-0.4, 3],
        random_seed=2021,
    )

    xys_list[1] = create_line_points(
        start=np.array([[4, 8.9]]),
        direction=direction,
        t_range=np.arange(-50, 50, 1.0),
        noise_range=[-1.0, 1.5],
        random_seed=2022,
    )
    return xys_list


def main():
    xys_list = create_data()
    max_iterations = 100
    sample_size = int(min(len(xys_list[0]), len(xys_list[1])) * 0.3)
    inlier_thresh = 0.1

    start = time.time()
    best_model, _ = run_ransac(
        xys_list,
        estimate,
        lambda xy, coeffs: is_inlier(xy, coeffs, inlier_thresh),
        sample_size,
        sample_size * 2,
        max_iterations,
        stop_at_goal=True,
    )
    n = best_model[:2]
    c = best_model[2:]
    print("n {}".format(n))
    print("c {}".format(c))

    print("\nruntime: {}[ms]\n".format((time.time() - start) * 1e3))

    colors = ["pink", "blue"]
    for (xys, color) in zip(xys_list, colors):
        plt.scatter(xys[:, 0], xys[:, 1], c=color)

    functions = [
        lambda x: (-c[0] - n[0] * x) / n[1],
        lambda x: (-c[1] - n[0] * x) / n[1],
    ]
    labels = ["first_line", "second_line"]
    for (xys, function, label) in zip(xys_list, functions, labels):
        plt.plot(xys[:, 0], function(xys[:, 0]), label=label)

    plt.legend()
    plt.show()


if __name__ == "__main__":
    main()

File: demo_parallel_line_fitting.jl
push!(LOAD_PATH, "$(@__DIR__)/src")
import LinearAlgebra
import Random
import Distributions
import PyPlot
import ParallelFitting


function create_line_points(;
    start::Array{Float64,2},
    direction::Array{Float64,2},
    t_range,
    noise_mean_std::Vector{Float64},
    random_seed::Int,
)::Matrix{Float64}
    @assert size(start) == (1, 2)
    @assert size(direction) == (1, 2)
    Random.seed!(random_seed)

    range_arr = hcat(t_range, t_range)
    points = start .+ (direction .* range_arr)
    noise = Random.rand(Distributions.Normal(noise_mean_std...), size(points))

    return points + noise
end


function create_data()::Array{Matrix{Float64}}
    direction = [3.2 6.7]
    direction /= LinearAlgebra.norm(direction) + 1e-10
    xys_list = Array{Matrix{Float64}}(undef, 2)
    xys_list[1] = create_line_points(;
        start = [20.0 10.0],
        direction = direction,
        t_range = 0:1.2:100,
        noise_mean_std = [-0.4, 3],
        random_seed = 2021,
    )
    xys_list[2] = create_line_points(
        start = [4 8.9],
        direction = direction,
        t_range = -50:1.0:50,
        noise_mean_std = [-1.0, 1.5],
        random_seed = 2022,
    )
    return xys_list
end


function main()
    xys_list = create_data()
    max_iterations = 100
    sample_size = floor(Int, min(size(xys_list[1])[1], size(xys_list[2])[1]) * 0.3)
    inlier_thresh = 0.5

    @time best_model, _ = ParallelFitting.run_ransac(
        xys_list = xys_list,
        estimate = ParallelFitting.Line.estimate,
        is_inlier = (xy, coeffs) -> ParallelFitting.Line.is_inlier(xy, coeffs, inlier_thresh),
        sample_size = sample_size,
        goal_inliers = sample_size * 2,
        max_iterations = max_iterations,
        stop_at_goal = true,
        random_seed = 2021,
    )

    n = best_model[begin:2]
    c = best_model[3:end]
    println("n $n")
    println("c $c")

    colors = ["pink", "blue"]
    for (xys, color) in zip(xys_list, colors)
        PyPlot.scatter(xys[:, 1], xys[:, 2], c = color)
    end

    functions = [x -> (-c[1] - n[1] * x) / n[2], x -> (-c[2] - n[1] * x) / n[2]]
    labels = ["first_line", "second_line"]

    for (xys, func, label) in zip(xys_list, functions, labels)
        PyPlot.plot(xys[:, 1], func.(xys[:, 1]), label = label)
    end

    PyPlot.show()
end


main()
